features/terrain.py

```diff
@@ -0,0 +1,253 @@
+
+            self._config[pg_name] = self._make_patroni_test_config(pg_name, self._output_dir)
+
+            p = subprocess.Popen(['python', 'patroni.py', self._config[pg_name]],
+                                 stdout=self._log[pg_name], stderr=subprocess.STDOUT, cwd=cwd)
+            if not (p and p.pid and p.poll() is None):
+                assert False, "PostgreSQL {0} is not running after being started".format(pg_name)
+            self._processes[pg_name] = p
+        # wait while patroni is available for queries, but not more than 10 seconds.
+        for tick in range(max_wait_limit):
+            if self.query(pg_name, "SELECT 1", fail_ok=True) is not None:
+                break
+            time.sleep(1)
+        else:
+            assert False,\
+                "Patroni instance is not available for queries after {0} seconds".format(max_wait_limit)
+
+    def stop(self, pg_name, kill=False):
+        while self._is_running(pg_name):
+            if not kill:
+                self._processes[pg_name].terminate()
+            else:
+                self._processes[pg_name].kill()
+            time.sleep(1)
+        self._log.get('pg_name') and self._log[pg_name].close()
+        if pg_name in self._processes:
+            del self._processes[pg_name]
+
+    def query(self, pg_name, query, fail_ok=False):
+        try:
+            cursor = self._cursor(pg_name)
+            cursor.execute(query)
+            return cursor
+        except psycopg2.Error:
+            if fail_ok:
+                return None
+            else:
+                raise
+
+    def check_role_has_changed_to(self, pg_name, new_role, timeout=10):
+        bound_time = time.time() + timeout
+        recovery_status = False if new_role == 'primary' else True
+        role_has_changed = False
+        while not role_has_changed:
+            cur = self.query(pg_name, "SELECT pg_is_in_recovery()", fail_ok=True)
+            if cur:
+                row = cur.fetchone()
+                if row and len(row) > 0 and row[0] == recovery_status:
+                    role_has_changed = True
+            if time.time() > bound_time:
+                break
+            time.sleep(1)
+        return role_has_changed
+
+    def stop_all(self):
+        for patroni in self._processes.copy():
+            self.stop(patroni)
+
+    def create_and_set_output_directory(self, feature_name):
+        feature_dir = os.path.join(pctl.patroni_path, "features", "output", feature_name.encode('utf-8').replace(' ', '_'))
+        if os.path.exists(feature_dir):
+            shutil.rmtree(feature_dir)
+        os.makedirs(feature_dir)
+        self._output_dir = feature_dir
+
+    def _is_running(self, pg_name):
+        return pg_name in self._processes and self._processes[pg_name].pid and (self._processes[pg_name].poll() is None)
+
+    def _make_patroni_test_config(self, pg_name, output_dir):
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Method could be a function](https://www.codacy.com/app/cyberdemn/patroni/file/2114827817/issues/source?bid=3127594&fileBranchId=3144623#l112)